torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -111,13 +123,15 @@ at::Tensor& AtenXlaType::__ilshift__(at::Tensor& self, at::Scalar other) {
 at::Tensor& AtenXlaType::__ilshift__(at::Tensor& self,
                                      const at::Tensor& other) {
   XLA_FN_COUNTER("xla::");
+  CheckBinaryOpTypePromotion(self, self, other);
```

Comment:
This is not right. There is no promotion going on in logical shifts.

Reply:
For ilshift, it is being considered as a binary op and type promotion is performed by pytorch. Here is the pytorch code 
 https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BinaryOps.cpp#L401
```
auto iter = TensorIterator::binary_op(self, self, other);
```
will call compute_types() and do the type promotion and check the promoted type with the self type. Here is an example

```
import torch

t1 = torch.tensor(1)
t2 = torch.tensor(1.5)

t1 <<= t2
```
and the result is 
```
(pytorch) jackcao@xujiacao:~/Desktop/pytorch/xla/test/cpp$ python ~/Desktop/toys/shift.py
Traceback (most recent call last):
  File "/usr/local/google/home/jackcao/Desktop/toys/shift.py", line 6, in <module>
    t1 <<= t2
RuntimeError: result type Float can't be cast to the desired output type Long
```

I think you are right that if we are not doing the type promotion for the __shift__ we should not do the type promotion check for the __ilshift__. I see two ways of doing this. First is I remove all of the type promotion check if we don't do the type promotion for the non-inplace version of the op. Second is I add the type promotion for all of the binary op. Which one do you think is more reasonable?

Reply:
Argh, yes. This is the generic, 2-power based, shift.

Reply:
Since we don't do type promotion for shift(by calling GetPromotedXlaTensorsForBinaryOp), should I keep this check here? Or is it better to do the type promotion for all of the binary ops? We currently only perform the type promotion for binary op 'add', 'mul', 'sub' and 'div'.

Reply:
Ouch ... that API is wrong (`GetPromotedXlaTensorsForBinaryOp()`).
It keeps changing the dtype of **source** tensors, so the tensor dtype changes multiple times along its life, according to the operation it is involved with.
```Python
Tensor i(Int)
Tensor l(Long)
z = i + l # i.dtype becomes Long after this op
```

Reply:
I am fixing the above API.
As far as promoting the remaining ones, we have to be consistent, so once the fix for the above API is in, you should do it.

Reply:
https://github.com/pytorch/xla/pull/1733

Reply:
Should we have a scalar version of doBinaryOp? like
```DoBinaryOp(const at::Tensor& self, const at::Scalar& other, const B& bin_op)```
I am calling CheckBinaryOpTypePromotion for the scalar version of the binary operations. Although it seems like xla is doing its own type promotion there but it might be better if we follow the same type promotion logic (by calling at::result_type) for these scalar binary ops as well.